File: adsResearchGlobal/src/config/sidebarNav.js
/**
 * Sidebar navigation structure.
 * Items support: requiresFullAccess, requiresProjectAccess, employeeOnly (hidden from full-access-only sections when false - N/A)
 */
export const getSidebarNav = ({ fullAccess, canViewProjects, allowedSections, dashboardPath = '/dashboard' }) => {
  const useSectionFilter = Array.isArray(allowedSections) && allowedSections.length > 0 && !fullAccess
  const isSectionAllowed = (section) => !useSectionFilter || allowedSections.includes(section.id)

  const filterItem = (item) => {
    if (item.requiresFullAccess && !fullAccess) return false
    if (item.requiresProjectAccess && !canViewProjects) return false
    return true
  }

  const filterGroup = (group) => {
    if (group.requiresFullAccess && !fullAccess) return null
    const children = (group.children || []).filter(filterItem)
    if (!children.length) return null
    return { ...group, children }
  }

  const sections = [
    {
      id: 'dashboard',
      label: 'Dashboard',
      icon: '🏠',
      type: 'link',
      path: dashboardPath,
    },
    {
      id: 'crm',
      label: 'CRM',
      icon: '👥',
      type: 'group',
      children: [
        { id: 'leads', label: 'Leads', path: '/leads' },
        { id: 'contacts', label: 'Contacts', path: '/collaborators', requiresFullAccess: true },
        { id: 'companies', label: 'Companies', path: '/companies', requiresFullAccess: true },
        { id: 'deals', label: 'Deals', path: '/module/deals', requiresFullAccess: true },
        { id: 'quotations', label: 'Quotations', path: '/module/quotations', requiresFullAccess: true },
      ],
    },
    {
      id: 'projects',
      label: 'Projects',
      icon: '📁',
      type: 'group',
      children: [
        { id: 'projects-list', label: 'Projects', path: '/projects', requiresProjectAccess: true },
        { id: 'tasks', label: 'Tasks', path: '/tasks', requiresFullAccess: true },
        { id: 'milestones', label: 'Milestones', path: '/module/milestones', requiresFullAccess: true },
        { id: 'timesheets', label: 'Timesheets', path: '/module/timesheets' },
      ],
    },
    {
      id: 'employees',
      label: 'Employees',
      icon: '👨',
      type: 'group',
      children: [
        { id: 'directory', label: 'Directory', path: '/employees', requiresFullAccess: true },
        { id: 'attendance', label: 'Attendance', path: '/attendance' },
        { id: 'leave', label: 'Leave', path: '/leave' },
        { id: 'performance', label: 'Performance', path: '/module/performance', requiresFullAccess: true },
        { id: 'assets', label: 'Assets', path: '/module/assets', requiresFullAccess: true },
      ],
    },
    {
      id: 'finance',
      label: 'Finance',
      icon: '💰',
      type: 'group',
      requiresFullAccess: true,
      children: [
        { id: 'invoices', label: 'Invoices', path: '/billings' },
        { id: 'expenses', label: 'Expenses', path: '/expenses' },
        { id: 'payments', label: 'Payments', path: '/revenue' },
        { id: 'payroll', label: 'Payroll', path: '/salaries' },
        { id: 'gst', label: 'GST / Taxes', path: '/module/gst' },
      ],
    },
    {
      id: 'inventory',
      label: 'Inventory',
      icon: '📦',
      type: 'group',
      requiresFullAccess: true,
      children: [
        { id: 'products', label: 'Products', path: '/module/products' },
        { id: 'stock', label: 'Stock', path: '/module/stock' },
        { id: 'purchase-orders', label: 'Purchase Orders', path: '/module/purchase-orders' },
        { id: 'suppliers', label: 'Suppliers', path: '/module/suppliers' },
      ],
    },
    {
      id: 'sales',
      label: 'Sales',
      icon: '🛒',
      type: 'group',
      children: [
        { id: 'orders', label: 'Orders', path: '/module/orders', requiresFullAccess: true },
        { id: 'customers', label: 'Customers', path: '/clients', requiresFullAccess: true },
        { id: 'pipeline', label: 'Sales Pipeline', path: '/lead-management' },
      ],
    },
    {
      id: 'support',
      label: 'Support',
      icon: '🛠',
      type: 'group',
      requiresFullAccess: true,
      children: [
        { id: 'tickets', label: 'Tickets', path: '/module/tickets' },
        { id: 'knowledge-base', label: 'Knowledge Base', path: '/module/knowledge-base' },
        { id: 'customer-feedback', label: 'Customer Feedback', path: '/module/customer-feedback' },
      ],
    },
    {
      id: 'marketing',
      label: 'Marketing',
      icon: '📢',
      type: 'group',
      children: [
        { id: 'campaigns', label: 'Campaigns', path: '/campaigns', requiresFullAccess: true },
        { id: 'email', label: 'Email', path: '/module/email', requiresFullAccess: true },
        { id: 'sms', label: 'SMS', path: '/module/sms', requiresFullAccess: true },
        { id: 'whatsapp', label: 'WhatsApp', path: '/module/whatsapp', requiresFullAccess: true },
        { id: 'social', label: 'Social Media', path: '/social-calendar' },
      ],
    },
    {
      id: 'documents',
      label: 'Documents',
      icon: '📄',
      type: 'group',
      requiresFullAccess: true,
      children: [
        { id: 'files', label: 'Files', path: '/module/files' },
        { id: 'contracts', label: 'Contracts', path: '/module/contracts' },
        { id: 'policies', label: 'Policies', path: '/module/policies' },
      ],
    },
    {
      id: 'reports',
      label: 'Reports & Analytics',
      icon: '📊',
      type: 'link',
      path: '/reports',
      requiresFullAccess: true,
    },
    {
      id: 'approvals',
      label: 'Approvals',
      icon: '✅',
      type: 'link',
      path: '/module/approvals',
      requiresFullAccess: true,
    },
    {
      id: 'communication',
      label: 'Communication',
      icon: '💬',
      type: 'group',
      children: [
        { id: 'chat', label: 'Chat', path: '/module/chat' },
        { id: 'announcements', label: 'Announcements', path: '/module/announcements' },
      ],
    },
    {
      id: 'administration',
      label: 'Administration',
      icon: '⚙',
      type: 'group',
      requiresFullAccess: true,
      children: [
        { id: 'users', label: 'Users', path: '/employees' },
        { id: 'roles', label: 'Roles & Permissions', path: '/module/roles' },
        { id: 'departments', label: 'Departments', path: '/module/departments' },
        { id: 'designations', label: 'Designations', path: '/module/designations' },
        { id: 'branches', label: 'Branches', path: '/module/branches' },
        { id: 'settings', label: 'Settings', path: '/settings' },
        { id: 'audit-logs', label: 'Audit Logs', path: '/module/audit-logs' },
      ],
    },
    {
      id: 'workspace',
      label: 'My Workspace',
      icon: '👤',
      type: 'group',
      children: [
        { id: 'my-profile', label: 'My Profile', path: '/my-profile' },
        { id: 'my-tasks', label: 'My Tasks', path: '/my-tasks' },
        { id: 'my-calendar', label: 'My Calendar', path: '/calendar' },
        { id: 'my-leaves', label: 'My Leaves', path: '/leave' },
        { id: 'my-attendance', label: 'My Attendance', path: '/attendance' },
        { id: 'my-projects', label: 'My Projects', path: '/my-projects' },
      ],
    },
  ]

  return sections
    .map((section) => {
      if (section.type === 'link') {
        if (section.requiresFullAccess && !fullAccess) return null
        if (!isSectionAllowed(section)) return null
        return section
      }
      const filtered = filterGroup(section)
      if (!filtered || !isSectionAllowed(filtered)) return null
      return filtered
    })
    .filter(Boolean)
}

File: centralized_app/lib/navigation/sidebar_nav.dart
class SidebarChildItem {
  const SidebarChildItem({
    required this.id,
    required this.label,
    required this.path,
    this.requiresFullAccess = false,
    this.requiresProjectAccess = false,
  });

  final String id;
  final String label;
  final String path;
  final bool requiresFullAccess;
  final bool requiresProjectAccess;
}

class SidebarEntry {
  const SidebarEntry.link({
    required this.id,
    required this.label,
    required this.icon,
    required this.path,
    this.requiresFullAccess = false,
    this.alwaysVisible = false,
  })  : type = SidebarEntryType.link,
        children = const [];

  const SidebarEntry.group({
    required this.id,
    required this.label,
    required this.icon,
    required this.children,
    this.requiresFullAccess = false,
    this.alwaysVisible = false,
  })  : type = SidebarEntryType.group,
        path = null;

  final String id;
  final String label;
  final String icon;
  final SidebarEntryType type;
  final String? path;
  final List<SidebarChildItem> children;
  final bool requiresFullAccess;
  final bool alwaysVisible;
}

enum SidebarEntryType { link, group }

class SidebarNavContext {
  const SidebarNavContext({
    required this.fullAccess,
    required this.canViewProjects,
    required this.isTeamLeader,
    required this.dashboardPath,
    required this.includeProperties,
    this.allowedSections,
  });

  final bool fullAccess;
  final bool canViewProjects;
  final bool isTeamLeader;
  final String dashboardPath;
  final bool includeProperties;
  final List<String>? allowedSections;
}

/// Mirrors web `sidebarNav.js` — one config for all companies.
class SidebarNav {
  SidebarNav._();

  static List<SidebarEntry> build(SidebarNavContext ctx) {
    final useSectionFilter =
        ctx.allowedSections != null && ctx.allowedSections!.isNotEmpty && !ctx.fullAccess;
    bool sectionAllowed(SidebarEntry section) {
      if (section.alwaysVisible) return true;
      if (!useSectionFilter) return true;
      return ctx.allowedSections!.contains(section.id);
    }

    bool itemAllowed(SidebarChildItem item) {
      if (item.requiresFullAccess && !ctx.fullAccess) return false;
      if (item.requiresProjectAccess && !ctx.canViewProjects) return false;
      return true;
    }

    SidebarEntry? filterGroup(SidebarEntry group) {
      if (group.requiresFullAccess && !ctx.fullAccess) return null;
      final children = group.children.where(itemAllowed).toList();
      if (children.isEmpty) return null;
      return SidebarEntry.group(
        id: group.id,
        label: group.label,
        icon: group.icon,
        requiresFullAccess: group.requiresFullAccess,
        alwaysVisible: group.alwaysVisible,
        children: children,
      );
    }

    final sections = <SidebarEntry>[
      SidebarEntry.link(
        id: 'dashboard',
        label: 'Dashboard',
        icon: '🏠',
        path: ctx.dashboardPath,
        alwaysVisible: true,
      ),
      if (ctx.isTeamLeader)
        SidebarEntry.group(
          id: 'my-team',
          label: 'My Team',
          icon: '👥',
          alwaysVisible: true,
          children: const [
            SidebarChildItem(id: 'team-members', label: 'Team Members', path: '/my-team'),
            SidebarChildItem(id: 'team-leave', label: 'Leave', path: '/leave'),
          ],
        ),
      SidebarEntry.group(
        id: 'workspace',
        label: 'My Workspace',
        icon: '👤',
        alwaysVisible: true,
        children: const [
          SidebarChildItem(id: 'my-profile', label: 'My Profile', path: '/my-profile'),
          SidebarChildItem(id: 'my-tasks', label: 'My Tasks', path: '/my-tasks'),
          SidebarChildItem(id: 'assign-task', label: 'Assign Task', path: '/assign-task'),
          SidebarChildItem(id: 'my-calendar', label: 'My Calendar', path: '/calendar'),
          SidebarChildItem(id: 'my-leaves', label: 'My Leaves', path: '/leave'),
          SidebarChildItem(id: 'my-attendance', label: 'My Attendance', path: '/my-attendance'),
          SidebarChildItem(id: 'my-projects', label: 'My Projects', path: '/my-projects'),
        ],
      ),
      SidebarEntry.group(
        id: 'crm',
        label: 'CRM',
        icon: '🎯',
        children: const [
          SidebarChildItem(id: 'leads', label: 'Leads', path: '/leads'),
          SidebarChildItem(
            id: 'contacts',
            label: 'Contacts',
            path: '/collaborators',
            requiresFullAccess: true,
          ),
          SidebarChildItem(
            id: 'companies',
            label: 'Companies',
            path: '/companies',
            requiresFullAccess: true,
          ),
          SidebarChildItem(
            id: 'quotations',
            label: 'Quotations',
            path: '/quotations',
            requiresFullAccess: true,
          ),
        ],
      ),
      if (ctx.includeProperties)
        SidebarEntry.group(
          id: 'properties',
          label: 'Property Listing',
          icon: '🏘',
          children: const [
            SidebarChildItem(id: 'property-listings', label: 'All Properties', path: '/properties'),
            SidebarChildItem(id: 'add-property', label: 'Add Property', path: '/add-property'),
            SidebarChildItem(id: 'site-visits', label: 'Site Visits', path: '/site-visits'),
            SidebarChildItem(
              id: 'schedule-site-visit',
              label: 'Schedule Visit',
              path: '/schedule-site-visit',
            ),
          ],
        ),
      SidebarEntry.group(
        id: 'projects',
        label: 'Projects',
        icon: '📁',
        children: const [
          SidebarChildItem(
            id: 'projects-list',
            label: 'Projects',
            path: '/projects',
            requiresProjectAccess: true,
          ),
          SidebarChildItem(id: 'tasks', label: 'Tasks', path: '/tasks', requiresFullAccess: true),
          SidebarChildItem(
            id: 'milestones',
            label: 'Milestones',
            path: '/module/milestones',
            requiresFullAccess: true,
          ),
          SidebarChildItem(id: 'timesheets', label: 'Timesheets', path: '/module/timesheets'),
        ],
      ),
      if (!ctx.isTeamLeader)
        SidebarEntry.group(
          id: 'employees',
          label: 'Employees',
          icon: '👨',
          children: const [
            SidebarChildItem(
              id: 'directory',
              label: 'Directory',
              path: '/employees',
              requiresFullAccess: true,
            ),
            SidebarChildItem(id: 'attendance', label: 'Attendance', path: '/attendance'),
            SidebarChildItem(id: 'leave', label: 'Leave', path: '/leave'),
            SidebarChildItem(
              id: 'performance',
              label: 'Performance',
              path: '/module/performance',
              requiresFullAccess: true,
            ),
            SidebarChildItem(
              id: 'assets',
              label: 'Assets',
              path: '/module/assets',
              requiresFullAccess: true,
            ),
          ],
        ),
      SidebarEntry.group(
        id: 'finance',
        label: 'Finance',
        icon: '💰',
        requiresFullAccess: true,
        children: const [
          SidebarChildItem(id: 'invoices', label: 'Invoices', path: '/billings'),
          SidebarChildItem(id: 'expenses', label: 'Expenses', path: '/expenses'),
          SidebarChildItem(id: 'payments', label: 'Revenue', path: '/revenue'),
          SidebarChildItem(id: 'payroll', label: 'Payroll', path: '/salaries'),
          SidebarChildItem(id: 'gst', label: 'GST / Taxes', path: '/module/gst'),
        ],
      ),
      SidebarEntry.group(
        id: 'sales',
        label: 'Sales',
        icon: '🛒',
        children: const [
          SidebarChildItem(id: 'orders', label: 'Orders', path: '/module/orders', requiresFullAccess: true),
          SidebarChildItem(id: 'customers', label: 'Customers', path: '/clients', requiresFullAccess: true),
          SidebarChildItem(id: 'pipeline', label: 'Sales Pipeline', path: '/lead-management'),
        ],
      ),
      SidebarEntry.group(
        id: 'marketing',
        label: 'Marketing',
        icon: '📢',
        children: const [
          SidebarChildItem(id: 'campaigns', label: 'Campaigns', path: '/campaigns', requiresFullAccess: true),
          SidebarChildItem(id: 'email', label: 'Email', path: '/module/email', requiresFullAccess: true),
          SidebarChildItem(id: 'sms', label: 'SMS', path: '/module/sms', requiresFullAccess: true),
          SidebarChildItem(id: 'whatsapp', label: 'WhatsApp', path: '/module/whatsapp', requiresFullAccess: true),
          SidebarChildItem(id: 'social', label: 'Social Media', path: '/social-calendar'),
        ],
      ),
      SidebarEntry.group(
        id: 'communication',
        label: 'Communication',
        icon: '💬',
        children: const [
          SidebarChildItem(id: 'chat', label: 'Chat', path: '/module/chat'),
          SidebarChildItem(id: 'announcements', label: 'Announcements', path: '/module/announcements'),
        ],
      ),
      SidebarEntry.link(
        id: 'reports',
        label: 'Reports & Analytics',
        icon: '📊',
        path: '/reports',
        requiresFullAccess: true,
      ),
      SidebarEntry.group(
        id: 'administration',
        label: 'Administration',
        icon: '⚙',
        requiresFullAccess: true,
        children: const [
          SidebarChildItem(id: 'users', label: 'Users', path: '/employees'),
          SidebarChildItem(id: 'departments', label: 'Departments', path: '/module/departments'),
          SidebarChildItem(id: 'designations', label: 'Designations', path: '/module/designations'),
          SidebarChildItem(id: 'settings', label: 'Settings', path: '/settings'),
        ],
      ),
    ];

    final out = <SidebarEntry>[];
    for (final section in sections) {
      if (ctx.isTeamLeader && section.id == 'employees') continue;
      if (section.type == SidebarEntryType.link) {
        if (section.requiresFullAccess && !ctx.fullAccess) continue;
        if (!sectionAllowed(section)) continue;
        out.add(section);
        continue;
      }
      final filtered = filterGroup(section);
      if (filtered == null) continue;
      if (!sectionAllowed(filtered)) continue;
      out.add(filtered);
    }
    return out;
  }

  static String? labelForPath(List<SidebarEntry> nav, String path) {
    for (final entry in nav) {
      if (entry.type == SidebarEntryType.link && entry.path == path) {
        return entry.label;
      }
      for (final child in entry.children) {
        if (child.path == path) return child.label;
      }
    }
    return null;
  }
}
